src/app/clusters/software-diagnostics-server/software-diagnostics-logic.cpp

```diff
@@ -0,0 +1,126 @@
+#include <lib/support/CodeUtils.h>
+
+#include <clusters/SoftwareDiagnostics/Commands.h>
+#include <clusters/SoftwareDiagnostics/Ids.h>
+#include <clusters/SoftwareDiagnostics/Metadata.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace {
+
+/// Wrapper around `DeviceLayer::GetDiagnosticDataProvider::GetThreadMetrics` that ensures
+/// that `ReleaseThreadMetrics` is always called on the underlying value.
+class AutoFreeThreadMetrics
+{
+public:
+    AutoFreeThreadMetrics(DeviceLayer::DiagnosticDataProvider & provider) : mProvider(provider) {}
+    ~AutoFreeThreadMetrics()
+    {
+        if (mMetrics != nullptr)
+        {
+            mProvider.ReleaseThreadMetrics(mMetrics);
+        }
+    }
+
+    CHIP_ERROR ReadThreadMetrics()
+    {
+        VerifyOrReturnError(mMetrics == nullptr, CHIP_ERROR_INCORRECT_STATE);
+        return mProvider.GetThreadMetrics(&mMetrics);
+    }
+
+    const DeviceLayer::ThreadMetrics * ThreadMetrics() const { return mMetrics; }
+
+private:
+    DeviceLayer::ThreadMetrics * mMetrics = nullptr;
+    DeviceLayer::DiagnosticDataProvider & mProvider;
+};
+
+} // namespace
+
+using namespace chip::app::Clusters::SoftwareDiagnostics;
+
+CHIP_ERROR SoftwareDiagnosticsLogic::ReadThreadMetrics(AttributeValueEncoder & encoder) const
+{
+    AutoFreeThreadMetrics metrics(GetDiagnosticDataProvider());
+
+    ReturnErrorOnFailure(metrics.ReadThreadMetrics());
+
+    return encoder.EncodeList([&metrics](const auto & itemEncoder) -> CHIP_ERROR {
+        for (const DeviceLayer::ThreadMetrics * thread = metrics.ThreadMetrics(); thread != nullptr; thread = thread->Next)
+        {
+            ReturnErrorOnFailure(itemEncoder.Encode(*thread));
+        }
+        return CHIP_NO_ERROR;
+    });
+}
+
+CHIP_ERROR SoftwareDiagnosticsLogic::AcceptedCommands(ReadOnlyBufferBuilder<DataModel::AcceptedCommandEntry> & builder)
+{
+    if (GetDiagnosticDataProvider().SupportsWatermarks())
+    {
+        static constexpr DataModel::AcceptedCommandEntry kAcceptedCommands[] = { Commands::ResetWatermarks::kMetadataEntry };
+        return builder.ReferenceExisting(kAcceptedCommands);
+    }
+
+    // no commands supported
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SoftwareDiagnosticsLogic::Attributes(ReadOnlyBufferBuilder<DataModel::AttributeEntry> & builder)
+{
+    // ensure we have space for all attributes. We may add at most 4 attributes (which are ALL optional or
+    // guarded by feature maps)
+    ReturnErrorOnFailure(builder.EnsureAppendCapacity(4 + DefaultServerCluster::GlobalAttributes().size()));
+
+    // NOTE:
+    //   - logic here is slow, however it is accurate.
+    //   - we COULD make the diagnostics provider give us a bitmap of support (maybe easiest however it changes API)
+    //     or cache this (would require a tiny amount of RAM).
+    AutoFreeThreadMetrics metrics(GetDiagnosticDataProvider());
+    if (metrics.ReadThreadMetrics() != CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE)
```

Comment:
I am not sure how I feel about tying support directly to the data provider and not being able to control whether the data is exposed if the provider claims to expose it.

Not least because absolutely nothing on the provider API documents that you return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE if you don't support it (or what the implications of that are), so there is zero guarantee (and imo low chances) that providers are doing that "correctly".